lib/elixir/lib/agent.ex

```diff
@@ -18,37 +18,27 @@ defmodule Agent do
       defmodule Mix.TasksServer do
         use Agent
 
-        def start_link(_) do
-          Agent.start_link(fn -> MapSet.new() end, name: __MODULE__)
+        def start_link(_opts) do
+          Agent.start_link(fn -> %{} end, name: __MODULE__)
         end
 
-        @doc "Checks if the task has already executed"
-        def executed?(task, project) do
-          item = {task, project}
-
-          Agent.get(__MODULE__, fn set ->
-            item in set
-          end)
+        def clear() do
+          update(fn _ -> %{} end)
         end
 
-        @doc "Marks a task as executed"
-        def put_task(task, project) do
-          item = {task, project}
-          Agent.update(__MODULE__, &MapSet.put(&1, item))
+        def put(tuple) do
+          update(&Map.put(&1, tuple, true))
         end
 
-        @doc "Resets the executed tasks and returns the previous list of tasks"
-        def take_all() do
-          Agent.get_and_update(__MODULE__, fn set ->
-            {Enum.into(set, []), MapSet.new()}
-          end)
+        defp update(fun) do
+          Agent.update(__MODULE__, fun, 30000)
```

Comment:
I'm actually not sure I agree with having an example like that in the documentation.

I'd say allowing passing arbitrary functions to the agent from outside leaks some implementation details - I think only having a well defined set of operations exposed as functions is a better approach. This is similar to having direct `GenServer.call/cast` calls in modules outside of the server itself.

Reply:
@michalmuskala this is a private function though? I often like to do exactly this, to encapsulate the agent name+timeout in a single place.

Reply:
Oh, I'm sorry. For some reason I missed the `p`. This looks fine now.

I guess it's just a little weird because we have write functions on the agent, but not a single read function.

Reply:
Good point. What if we change the example with a very simple `get`/`set` agent?

Reply:
The "reading" function in `Mix.TasksServer` is

```elixir
  def run(tuple) do
    get_and_update(fn set ->
      {not Map.has_key?(set, tuple), Map.put(set, tuple, true)}
    end)
  end
```

But without more context, I guess a function called `run` that doesn't execute anything may be confusing.

The KV example in https://elixir-lang.org/getting-started/mix-otp/agent.html, for example, seems more simple to me. It could be even an agent that stores a single scalar like a number or something like that (but still with a realistic use case).

Reply:
One possibility could be an agent that provides a synchronized sequence within a node:

```elixir
defmodule Sequence do
  use Agent

  def start_link do
    Agent.start_link(fn -> 0 end, name: __MODULE__)
  end

  def next do
    Agent.get_and_update(__MODULE__, &({&1, &1 + 1}))
  end
end
```

As first useful agent example, it can't get more simple than that! 😄

Reply:
Yes, I think either a counter or a get-set register would be good examples, that don't require the reader to also learn the problem, but just the solution.